CTCP Fixes

andymandias opened the pull request:
Fixes for #428, #429, and #431.  Also adds rich text parsing to CTCP `ACTION`s.

glguy reviewed (approved):
everything seems to be in order!

the client doesn't show CTCP queries that aren't actions when sent to channel. that's probably intentional but for a channel operator it's not great to not know what's being sent to your channel. If you think that's an issue that's a topic for a different PR, I think.

glguy commented:
Actually, I found a problem. The PING reply doesn't send the PING command; it just sends the parameters directly without the PING

andymandias commented:
Thank you for testing!  I really should find a secondary client to use for testing purposes.

I don't have any opinion on non-actions sent to channels.

glguy commented:
It's normal to have two copies of CLIENTINFO. The first is the response command and the second indicates that that is a supported command. This duplication shows up in the standard.

andymandias commented:
That was the one command where I had added the command name to response, so when I added another we had three `CLIENTINFO`s.  I alphabetized the response list, so the two `CLIENTINFO`s don't show up next to each other.

tarkah commented on data/src/ctcp.rs:
```diff
@@ -15,10 +15,13 @@ pub fn parse_query(text: &str) -> Option<Query> {
+            command: command.to_uppercase(),
+            params,
+        })
     } else {
         Some(Query {
-            command: query,
+            command: query.to_uppercase(),
```
We are relying on an internal implementation detail from another module, to know we should match on this "uppercase".

Instead, let's define a Command enum in this module and match on this uppercase guarantee _internally_ to produce the strongly typed enum. `client` module can then rely on the strongly typed Command guarantee, making this impossible to misuse.

tarkah commented on data/src/client.rs:
```diff
@@ -505,42 +505,44 @@ impl Client {
                                         ));
+                                        }
+                                        "PING" => {
+                                            let _ = self.handle.try_send(command!(
+                                                "NOTICE",
+                                                user.nickname().to_string(),
+                                                format!("\u{1}PING {}\u{1}", query.params)
+                                            ));
```
I smell a `ctcp::format` method which takes a `ctcp::Command`, `User` and `&str` / params and formats these IRC messages so we don't have to define it here and accidentally miss `\u{1}` or like before, missing the `COMMAND` in the response

andymandias commented:
Thanks for the review and the instructive suggestions (as usual) @tarkah.  I think they're all implemented at this point.

tarkah reviewed (approved):
LGTM! Thanks for tackling these

andymandias commented:
@glguy if you want to make an issue for the re: showing CTCP queries sent to channels I can take a look at it when I've got some time.  Sounds like it might make sense to provide for operators if we gate it behind a setting.
